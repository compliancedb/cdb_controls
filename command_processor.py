from cdb.api_schema import ApiSchema
from cdb.http import http_put_payload


def execute(env):
    command = env.get("MERKELY_COMMAND", None)
    print("MERKELY_COMMAND={}".format(command))
    merkleypipe_path = "/Merkelypipe.json"
    with open(merkleypipe_path) as merkelypipe_file:
        merkelypipe = load_merkelypipe(merkelypipe_file)
        api_token = env.get('MERKELY_API_TOKEN', None)
        host = "https://app.compliancedb.com"
        if command == "declare_pipeline":
            declare_pipeline(env, merkelypipe, api_token, host)
        if command == "log_artifact":
            log_artifact(env, merkelypipe, api_token, host)

    return 0


def declare_pipeline(env, merkelypipe, api_token, host):
    pipelines_url = ApiSchema.url_for_pipelines(host, merkelypipe)
    http_put_payload(url=pipelines_url, payload=merkelypipe, api_token=api_token)


def log_artifact(env, merkelypipe, api_token, host):
    fingerprint = env.get("MERKELY_FINGERPRINT", None)
    protocol = fingerprint[:len("file://")]
    artifact_name = '/' + fingerprint[len("file://"):]
    print("Getting SHA for artifact: " + artifact_name)  # print "file" ?
    artifact_sha = sha_digest_for_file(artifact_name)
    print("Calculated digest: " + artifact_sha)
    #print("Publish artifact to ComplianceDB")
    description = "Created by build " + env.get('MERKELY_CI_BUILD_NUMBER', None)
    git_commit = env.get('MERKELY_ARTIFACT_GIT_COMMIT', None)
    commit_url = env.get('MERKELY_ARTIFACT_GIT_URL', None)
    build_url = env.get('MERKELY_CI_BUILD_URL', None)
    # is_compliant = env_is_compliant()
    # print('CDB_IS_COMPLIANT: ' + str(is_compliant))
    create_artifact(api_token, host, merkelypipe,
                    artifact_sha, artifact_name,
                    description, git_commit, commit_url, build_url)


import json
def load_merkelypipe(file):
    return json.load(file)


import subprocess
def sha_digest_for_file(artifact_path):
    output = subprocess.check_output(["openssl", "dgst", "-sha256", artifact_path])
    digest_in_bytes = output.split()[1]
    artifact_sha = digest_in_bytes.decode('utf-8')
    return artifact_sha


def create_artifact(api_token, host, merkelypipe,
                    sha256, filename,
                    description, git_commit, commit_url, build_url):
    create_artifact_payload = {
        "sha256": sha256,
        "filename": filename,
        "description": description,
        "git_commit": git_commit,
        "commit_url": commit_url,
        "build_url": build_url
        #"is_compliant": is_compliant
    }
    url = ApiSchema.url_for_artifacts(host, merkelypipe)
    http_put_payload(url, create_artifact_payload, api_token)
