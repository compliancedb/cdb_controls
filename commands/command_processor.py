from cdb.api_schema import ApiSchema
from cdb.http import http_put_payload
import json


def execute(context):
    command = context['env'].get("MERKELY_COMMAND", None)
    if command == "declare_pipeline":
        DeclarePipelineCommand(context).execute()
    if command == "log_artifact":
        LogArtifactCommand(context).execute()
    return 0


class Command:

    def __init__(self, context):
        self._context = context

    def execute(self):
        print("MERKELY_COMMAND={}".format(self.command))
        self.concrete_execute()

    @property
    def command(self):
        return self._env("MERKELY_COMMAND")

    @property
    def api_token(self):
        return self._env("MERKELY_API_TOKEN")

    @property
    def host(self):
        return self._env("MERKELY_HOST")

    @property
    def merkelypipe(self):
        MERKELYPIPE_PATH = "/Merkelypipe.json"
        with open(MERKELYPIPE_PATH) as file:
            return json.load(file)

    def _env(self, name):
        return self._context['env'].get(name, None)


class DeclarePipelineCommand(Command):

    def concrete_execute(self):
        pipelines_url = ApiSchema.url_for_pipelines(self.host, self.merkelypipe)
        http_put_payload(url=pipelines_url, payload=self.merkelypipe, api_token=self.api_token)


class LogArtifactCommand(Command):

    def concrete_execute(self):
        fingerprint = self._env("MERKELY_FINGERPRINT")
        context = self._context
        context['description'] = "Created by build " + self._env('MERKELY_CI_BUILD_NUMBER')
        context['git_commit'] = self._env('MERKELY_ARTIFACT_GIT_COMMIT')
        context['commit_url'] = self._env('MERKELY_ARTIFACT_GIT_URL')
        context['build_url'] = self._env('MERKELY_CI_BUILD_URL')
        context['is_compliant'] = self._env('MERKELY_IS_COMPLIANT') == "TRUE"

        FILE_PROTOCOL = "file://"
        if fingerprint.startswith(FILE_PROTOCOL):
            self._log_artifact_file(fingerprint[len(FILE_PROTOCOL):])

        DOCKER_PROTOCOL = "docker://"
        if fingerprint.startswith(DOCKER_PROTOCOL):
            self._log_artifact_docker_image(fingerprint[len(DOCKER_PROTOCOL):])

    def _log_artifact_file(self, filename):
        context = self._context
        pathed_filename = '/' + filename
        print("Getting SHA for file:// artifact: " + pathed_filename)
        artifact_sha = context['sha_digest_for_file'](pathed_filename)
        print("Calculated digest: " + artifact_sha)
        # print("Publish artifact to ComplianceDB")
        print('MERKELY_IS_COMPLIANT: ' + str(context['is_compliant']))
        create_artifact(self.api_token, self.host, self.merkelypipe,
                        artifact_sha, pathed_filename,
                        context['description'],
                        context['git_commit'],
                        context['commit_url'],
                        context['build_url'],
                        context['is_compliant'])

    def _log_artifact_docker_image(self, image_name):
        context = self._context
        print("Getting SHA for docker:// artifact: " + image_name)
        artifact_sha = context['sha_digest_for_docker_image'](image_name)
        print("Calculated digest: " + artifact_sha)
        # print("Publish artifact to ComplianceDB")
        print('MERKELY_IS_COMPLIANT: ' + str(context['is_compliant']))
        create_artifact(self.api_token, self.host, self.merkelypipe,
                        artifact_sha, image_name,
                        context['description'],
                        context['git_commit'],
                        context['commit_url'],
                        context['build_url'],
                        context['is_compliant'])



def create_artifact(api_token, host, merkelypipe,
                    sha256, filename,
                    description, git_commit, commit_url, build_url, is_compliant):
    create_artifact_payload = {
        "sha256": sha256,
        "filename": filename,
        "description": description,
        "git_commit": git_commit,
        "commit_url": commit_url,
        "build_url": build_url,
        "is_compliant": is_compliant
    }
    url = ApiSchema.url_for_artifacts(host, merkelypipe)
    http_put_payload(url, create_artifact_payload, api_token)
