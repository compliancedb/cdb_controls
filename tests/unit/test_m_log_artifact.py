import command_processor
from tests.utils import verify_approval, ScopedEnvVars, ScopedFileCopier, CDB_DRY_RUN


def test_command_processor_log_artifact_file(capsys):
    commit = "abc50c8a53f79974d615df335669b59fb56a4ed3"
    env = {
        "MERKELY_COMMAND": "log_artifact",
        "MERKELY_API_TOKEN": "MY_SUPER_SECRET_API_TOKEN",
        "MERKELY_FINGERPRINT": "file://coverage.txt",
        "MERKELY_CI_BUILD_URL": "https://gitlab/build/1456",
        "MERKELY_CI_BUILD_NUMBER": "23",
        "MERKELY_ARTIFACT_GIT_URL": "http://github/me/project/commit/" + commit,
        "MERKELY_ARTIFACT_GIT_COMMIT": commit,
    }

    with ScopedEnvVars({**CDB_DRY_RUN, **env}) as ev:
        with ScopedFileCopier("/app/tests/data/coverage.txt", "/coverage.txt"):
            with ScopedFileCopier("/app/tests/data/Merkelypipe.json", "/Merkelypipe.json"):
                assert ev.get("MERKELY_COMMAND") == "log_artifact"
                status_code = command_processor.execute(ev)

    assert status_code == 0
    verify_approval(capsys)

