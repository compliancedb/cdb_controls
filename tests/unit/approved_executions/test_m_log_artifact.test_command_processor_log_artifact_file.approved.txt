MERKELY_COMMAND=log_artifact
Getting SHA for artifact: /coverage.txt
Calculated digest: ccee89ccdc05772d90dc6929ad4f1fbc14aa105addf3326aa5cf575a104f51dc
Putting this payload:
{
    "build_url": "https://gitlab/build/1456",
    "commit_url": "http://github/me/project/commit/abc50c8a53f79974d615df335669b59fb56a4ed3",
    "description": "Created by build 23",
    "filename": "/coverage.txt",
    "git_commit": "abc50c8a53f79974d615df335669b59fb56a4ed3",
    "sha256": "ccee89ccdc05772d90dc6929ad4f1fbc14aa105addf3326aa5cf575a104f51dc"
}
To this url: https://app.compliancedb.com/api/v1/projects/merkely-test/merkely-change-test-pipeline/artifacts/
DRY RUN: Put not sent
